### PR TITLE
Codegen sqrt and fabs directly to llvm intrinsics when possible

### LINKED ISSUE #24112: sqrt and --no-ieee-float mystery on SLES 12
PRs #24106 and #24111 add a .suppressif for `check-sqrtr-noieee`.

After PR #24066, the new test `check-sqrt` was failing on some systems, but not others, when using `--no-ieee-float`. `--no-ieee-float` translates to `-ffast-math` and should enable more optimization. However, on the affected systems, compiling with `--no-ieee-float --fast` fails to inline the `sqrt` call into an instruction on that system.

<details>

```
$ chpl check-sqrt-noieee.chpl  --llvm-print-ir multiSqrt --llvm-print-ir-stage asm --fast --no-ieee-float

1 symbol names to disassemble


# Disassembling symbol multiSqrt_chpl


/tmp/chpl-mferguson.deleteme-K1RtoM/chpl__module.o:	file format elf64-x86-64

Disassembly of section .text:

00000000000301d0 <multiSqrt_chpl>:
   301d0: 50                           	pushq	%rax
   301d1: c5 f9 2e 05 00 00 00 00      	vucomisd	, %xmm0 <multiSqrt_chpl+0x9>
   301d9: 76 14                        	jbe	 <L0>
   301db: 0f 1f 44 00 00               	nopl	(%rax,%rax)
<L2>:
   301e0: e8 00 00 00 00               	callq	 <L1>
<L1>:
   301e5: c5 f9 2e 05 00 00 00 00      	vucomisd	, %xmm0 <multiSqrt_chpl+0x1d>
   301ed: 77 f1                        	ja	 <L2>
<L0>:
   301ef: 58                           	popq	%rax
   301f0: c3                           	retq
   301f1: 66 66 66 66 66 66 2e 0f 1f 84 00 00 00 00 00 	nopw	%cs:(%rax,%rax)
```

(note that the `callq` above is calling `sqrt`; we just don't see `sqrt` mentioned due to not using `llvm-objdump -r` due to https://github.com/llvm/llvm-project/issues/75539 )

</details>

At the same time, on the affected systems, compiling a similar C program with the `clang` we are configured to use shows the expected inlining:

<details>

``` c
// test.c
#include <math.h>
#include <stdio.h>

__attribute__((noinline))
static double multiSqrt(double n) {
  while (n > 10.0) {
    n = sqrt(n);
  }
  return n;
}

int main(int argc, char** argv) {
  double nums[1000];
  double sum = 0.0;
  for (int i = 0; i < 1000; i++) {
    nums[i] = multiSqrt((double)i);
  }

  for (int i = 0; i < 1000; i++) {
    sum += nums[i];
  }

  printf("sum is %f\n", sum);
  return 0;
}
```

```
$ /data/cf/chapel/chpl-deps/spack/opt/spack/linux-sles12-haswell/gcc-12.2.0/llvm-15.0.4-k67u3v54wddzwta5ecwljfvjh4lxkrnm/bin/clang test.c -O3 -ffast-math -fno-math-errno -S
$ cat test.s
	.text
	.file	"test.c"
	.globl	main                            # -- Begin function main
	.p2align	4, 0x90
	.type	main,@function

...

	.section	.rodata.cst8,"aM",@progbits,8
	.p2align	3                               # -- Begin function multiSqrt
.LCPI1_0:
	.quad	0x4024000000000000              # double 10
	.text
	.p2align	4, 0x90
	.type	multiSqrt,@function
multiSqrt:                              # @multiSqrt
	.cfi_startproc
# %bb.0:
	ucomisd	.LCPI1_0(%rip), %xmm0
	jbe	.LBB1_3
# %bb.1:
	movsd	.LCPI1_0(%rip), %xmm1           # xmm1 = mem[0],zero
	.p2align	4, 0x90
.LBB1_2:                                # =>This Inner Loop Header: Depth=1
	sqrtsd	%xmm0, %xmm0
	ucomisd	%xmm1, %xmm0
	ja	.LBB1_2
.LBB1_3:
	retq
.Lfunc_end1:
	.size	multiSqrt, .Lfunc_end1-multiSqrt
	.cfi_endproc
                                        # -- End function
...
```

Here we can see that there is no `callq` in the body of `multiSqrt`, as expected.

</details>

It is a mystery to me why this is occurring. The problem seems specific to OS version rather than processor type. In particular, all of the affected systems are running SUSE Linux Enterprise Server 12 SP3, which comes with GCC 4.8.5. It does not seem to be an issue on SUSE Linux Enterprise Server 15 SP3, which comes with GCC 7.5.0. Similarly, it does not seem to be an issue on Ubuntu 23.10 which comes with GCC 13.2.0. It is also possible that it has to do with how the system LLVM and clang were built on the affected systems.

### PR DESCRIPTION
**Pull request by jabraham17**
Adjusts the codegen for `sqrt` and `abs` to directly codegen to an llvm ir intrinsic when using the llvm backend (and not using strict floating point ops). On many systems, this will have no impact with `--fast` as the C library call previously used was optimized into this intrinsic. However, on older glibc versions this pr should allow more code to be vectorizable.

Testing
- [x] linux64
- [x] linux64 with gasnet
- [x] linux64 with older glibc

Resolves #24112

[Reviewed by @mppf]

### CODE REVIEW

File: modules/internal/ChapelBase.chpl [hunk 1]
```diff
@@ -3441,4 +3441,30 @@ module ChapelBase {
   inline proc chpl_field_gt(a, b) where !isArrayType(a.type) {
     return a > b;
   }
+
+  export inline proc chpl_sqrt64(x: real(64)): real(64) {
```
Comment (mppf) [inline modules/internal/ChapelBase.chpl:3445]: I would expect these to be in `runtime/include/chplmath.h` and just be called from the code generator. AFAIK you don't even have to declare them as `extern proc` in the Chapel code at all.

**Review by mppf — APPROVED** [review]

File: modules/internal/ChapelBase.chpl (hunk 1 above)
Comment (jabraham17) [reply modules/internal/ChapelBase.chpl]: Done in https://github.com/chapel-lang/chapel/pull/24455/commits/bec6eeac42644a39cb1940a23cae240c8cdc9d8f.

File: compiler/AST/primitive.cpp [hunk 2]
```diff
@@ -166,7 +166,7 @@ returnInfoAbs(CallExpr* call) {
     return QualifiedType(dtReal[FLOAT_SIZE_64], QUAL_VAL);
   }
 
-  return QualifiedType(t, QUAL_VAL);
+  return call->get(1)->qualType();
```
[the remark this answers is not in the document]
Comment (jabraham17) [reply compiler/AST/primitive.cpp:169]: I was getting very weird internal errors without this change, something in codegenAssign that only affected `abs`. I just reverted this change locally and tried again to compile `test/llvm/assertVectorize/math-calls.chpl` and could not replicate it. So I will just revert this.

**Comment by bradcray** [comment]
Thanks for this, @jabraham17!

**Pull request merged**